File: app/src/main/java/br/com/jonathanarodr/playmovie/ui/MovieAdapter.java
package br.com.jonathanarodr.playmovie.ui;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.squareup.picasso.Picasso;

import java.util.List;

import br.com.jonathanarodr.playmovie.R;
import br.com.jonathanarodr.playmovie.model.Movie;

public class MovieAdapter extends RecyclerView.Adapter<MovieAdapter.MovieViewHolder> {

    private final MovieAdapterOnClickHandler mClickHandler;
    private List<Movie> mMovies;

    public MovieAdapter(MovieAdapterOnClickHandler clickHandler) {
        mClickHandler = clickHandler;
    }

    public void setMovies(List<Movie> movies) {
        mMovies = movies;
        notifyDataSetChanged();
    }

    public List<Movie> getMovies() {
        return mMovies;
    }

    public class MovieViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private final ImageView mPoster;

        MovieViewHolder(View itemView) {
            super(itemView);
            mPoster = itemView.findViewById(R.id.poster);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            mClickHandler.onClick(mMovies.get(getAdapterPosition()));
        }

    }


    @NonNull
    @Override
    public MovieViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_movie, parent, false);
        return new MovieViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MovieViewHolder holder, int position) {
        Movie movie = mMovies.get(position);
        Picasso.get().load(movie.getPoster()).placeholder(R.drawable.backgroud_grey).into(holder.mPoster);
    }

    @Override
    public int getItemCount() {
        return (mMovies == null) ? 0 : mMovies.size();
    }

}

File: app/src/main/java/br/com/jonathanarodr/playmovie/ui/MovieAdapter.kt
package br.com.jonathanarodr.playmovie.ui

import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewGroup
import android.widget.ImageView
import androidx.recyclerview.widget.RecyclerView.Adapter
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import br.com.jonathanarodr.playmovie.R
import br.com.jonathanarodr.playmovie.R.id
import br.com.jonathanarodr.playmovie.model.Movie
import br.com.jonathanarodr.playmovie.util.GlideApp

class MovieAdapter(private val clickHandler: MovieAdapterOnClickHandler) :
    Adapter<MovieAdapter.MovieViewHolder>() {

    var movies: List<Movie> = emptyList()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    inner class MovieViewHolder(itemView: View) : ViewHolder(itemView), OnClickListener {
        val poster: ImageView = itemView.findViewById(id.poster)

        init {
            itemView.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            clickHandler.onClick(this@MovieAdapter.movies[adapterPosition])
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieViewHolder {
        val view: View = LayoutInflater
            .from(parent.context).inflate(R.layout.item_movie, parent, false)

        return MovieViewHolder(view)
    }

    override fun onBindViewHolder(holder: MovieViewHolder, position: Int) {
        this.movies[position].let {
            GlideApp.with(holder.itemView.context)
                .load(it.getPosterDefault())
                .centerCrop()
                .placeholder(R.drawable.backgroud_grey)
                .into(holder.poster)
        }
    }

    override fun getItemCount(): Int = movies.size
}